pkg/datapath/loader/cache.go

```diff
@@ -0,0 +1,235 @@
+	return newObjectCache(dp, nodeCfg, ".")
+}
+
+// Update may be called to update the base hash for configuration of datapath
+// configuration that applies across the node.
+func (o *ObjectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
+	newHash := HashDatapath(o.Datapath, nodeCfg, nil, nil)
+
+	o.Lock()
+	defer o.Unlock()
+	o.baseHash = newHash
+}
+
+// serialize finds the channel that serializes builds against the same hash.
+// Returns the channel and whether or not the caller needs to compile the
+// datapath for this hash.
+func (o *ObjectCache) serialize(hash string) (fq *serializer.FunctionQueue, found bool) {
+	o.Lock()
+	defer o.Unlock()
+
+	fq, compiled := o.compileQueue[hash]
+	if !compiled {
+		fq = serializer.NewFunctionQueue(1)
+		o.compileQueue[hash] = fq
+	}
+	return fq, compiled
+}
+
+func (o *ObjectCache) lookup(hash string) (string, bool) {
+	o.Lock()
+	defer o.Unlock()
+	path, exists := o.toPath[hash]
+	return path, exists
+}
+
+func (o *ObjectCache) insert(hash, objectPath string) {
+	o.Lock()
+	defer o.Unlock()
+	o.toPath[hash] = objectPath
+}
+
+// build attempts to compile and cache a datapath template object file
+// corresponding to the specified endpoint configuration.
+func (o *ObjectCache) build(ctx context.Context, cfg *templateCfg, hash string) error {
+	templatePath := filepath.Join(o.workingDirectory, defaults.TemplatesDir, hash)
+	headerPath := filepath.Join(templatePath, common.CHeaderFileName)
+	objectPath := filepath.Join(templatePath, endpointObj)
+
+	if err := os.MkdirAll(templatePath, defaults.StateDirRights); err != nil {
+		return err
+	}
+
+	f, err := os.Create(headerPath)
+	if err != nil {
+		return &os.PathError{
+			Op:   "failed to open template header for writing",
+			Path: headerPath,
+			Err:  err,
+		}
+	}
+
+	if err = o.Datapath.WriteEndpointConfig(f, cfg); err != nil {
+		return &os.PathError{
+			Op:   "failed to write template header",
+			Path: headerPath,
+			Err:  err,
+		}
+	}
+
+	cfg.stats.bpfCompilation.Start()
+	err = CompileTemplate(ctx, templatePath)
+	cfg.stats.bpfCompilation.End(err == nil)
+	if err != nil {
+		return err
+	}
+
+	log.WithFields(logrus.Fields{
+		logfields.Path:               objectPath,
+		logfields.BPFCompilationTime: cfg.stats.bpfCompilation.Total(),
+	}).Debug("Compiled new BPF template")
+
+	o.insert(hash, objectPath)
+	return nil
+}
+
+// FetchOrCompile attempts to fetch the path to the datapath object
+// corresponding to the provided endpoint configuration, or if this
+// configuration is not yet compiled, compiles it. It will block if multiple
+// threads attempt to concurrently FetchOrCompile a template binary for the
+// same set of EndpointConfiguration.
+//
+// Returns the path to the compiled template datapath object and whether the
+// object was compiled, or an error.
+func (o *ObjectCache) FetchOrCompile(ctx context.Context, cfg datapath.EndpointConfiguration, stats *SpanStat) (string, bool, error) {
```

Comment:
We actually end up taking the lock once to serialize here, then once more to look up the result once we know the template is compiled. This could be folded into one lock in the "template already exists" case. But that said there's probably not going to be a lot of contention on this, and it's not run super often either.